src/librustc/infer/nll_relate/mod.rs

```diff
@@ -616,15 +616,32 @@ where
     fn consts(
         &mut self,
         a: &'tcx ty::Const<'tcx>,
-        b: &'tcx ty::Const<'tcx>,
+        mut b: &'tcx ty::Const<'tcx>,
     ) -> RelateResult<'tcx, &'tcx ty::Const<'tcx>> {
-        if let ty::Const { val: ConstValue::Bound(..), .. } = a {
-            // FIXME(const_generics): I'm unsure how this branch should actually be handled,
-            // so this is probably not correct.
-            self.infcx.super_combine_consts(self, a, b)
-        } else {
-            debug!("consts(a={:?}, b={:?}, variance={:?})", a, b, self.ambient_variance);
-            relate::super_relate_consts(self, a, b)
+        let a = self.infcx.shallow_resolve(a);
+
+        if !D::forbid_inference_vars() {
+            b = self.infcx.shallow_resolve(b);
+        }
+
+        match (a.val, b.val) {
```

Comment:
We could probably replace this with:
```rust
match b.val {
	ConstValue::Infer(InferConst::Var(_)) if D::forbid_inference_vars() => {
		// Forbid inference variables in the RHS.
        bug!("unexpected inference var {:?}", b)
	}
	_ => self.infcx.super_combine_consts(self, a, b)
}
```
but I'm not sure if this obfuscates the correspondence with `tys` too much.